core/src/test/scala/org/apache/spark/storage/ShuffleBlockFetcherIteratorSuite.scala

```diff
@@ -123,6 +131,54 @@ class ShuffleBlockFetcherIteratorSuite extends SparkFunSuite with PrivateMethodT
     verify(wrappedInputStream.invokePrivate(delegateAccess()), times(1)).close()
   }
 
+  /**
+   * Get a blockByAddress iterator for a single BlockManagerId assuming all blocks have the same
+   * size and `blockMapId`.
+   */
+  private def getBlocksByAddressForSingleBM(
```

Comment:
Maybe this could be extended to support the case of multiple blockmanagers, e.g.,

```scala
val blocksByAddress = Seq[(BlockManagerId, Seq[(BlockId, Long, Int)])](
  (localBmId, localBlocks.keys.map(blockId => (blockId, 1L, 0)).toSeq),
  (remoteBmId, remoteBlocks.keys.map(blockId => (blockId, 1L, 1)).toSeq),
  (hostLocalBmId, hostLocalBlocks.keys.map(blockId => (blockId, 1L, 1)).toSeq)
).toIterator
```

We can pass in a `Map[BlockManagerId, (blocks, size, mapIndex)]` instead.

WDYT?

Reply:
Good idea! I focused on the single-BM case because it was simpler, but you're right that there was still a lot of common logic to be reduced. Actually, your suggestion made me realize that we only ever use this method (and the new multi-BM method I created) to create an iterator which is then passed to `getShuffleIteratorWithDefaults`, so I just made `getShuffleIteratorWithDefaults` directly accept a `Map`. I think it's very clean now, thank you for the suggestion!